evm/chains/sharding/mainchain_handler/vmc_utils.py

```diff
@@ -0,0 +1,212 @@
+
+viper_rlp_decoder_addr = eth_utils.to_checksum_address(
+    generate_contract_address(eth_utils.to_canonical_address(viper_rlp_decoder_tx.sender), 0)
+)
+
+sighasher_tx = rlp.decode(eth_utils.decode_hex("0xf9016d808506fc23ac0083026a508080b9015a6101488061000e6000396101565660007f01000000000000000000000000000000000000000000000000000000000000006000350460f8811215610038576001915061003f565b60f6810391505b508060005b368312156100c8577f01000000000000000000000000000000000000000000000000000000000000008335048391506080811215610087576001840193506100c2565b60b881121561009d57607f8103840193506100c1565b60c08112156100c05760b68103600185013560b783036020035260005101840193505b5b5b50610044565b81810360388112156100f4578060c00160005380836001378060010160002060e052602060e0f3610143565b61010081121561010557600161011b565b6201000081121561011757600261011a565b60035b5b8160005280601f038160f701815382856020378282600101018120610140526020610140f350505b505050505b6000f31b2d4f"), SpuriousDragonTransaction)
+sighasher_addr = eth_utils.to_checksum_address(
+    generate_contract_address(eth_utils.to_canonical_address(sighasher_tx.sender), 0)
+)
+
+def mk_validation_code(address):
+    '''
+    validation_code = """
+~calldatacopy(0, 0, 128)
+~call(3000, 1, 0, 0, 128, 0, 32)
+return(~mload(0) == {})
+    """.format(utils.checksum_encode(address))
+    return serpent.compile(validation_code)
+    '''
+    # The precompiled bytecode of the validation code which
+    # verifies EC signatures
+    validation_code_bytecode = b"a\x009\x80a\x00\x0e`\x009a\x00GV`\x80`\x00`\x007` "
+    validation_code_bytecode += b"`\x00`\x80`\x00`\x00`\x01a\x0b\xb8\xf1Ps"
+    validation_code_bytecode += address
+    validation_code_bytecode += b"`\x00Q\x14` R` ` \xf3[`\x00\xf3"
+    return validation_code_bytecode
+
+def sign(message, privkey):
+    '''@privkey: Key type
+    '''
+    keys = KeyAPI()
+    signature = keys.ecdsa_sign(message, privkey)
+    v, r, s = signature.vrs
+    v += 27
+    signature_bytes = b''.join([item.to_bytes(32, 'big') for item in (v, r, s)])
+    return signature_bytes
+
+def get_valmgr_abi():
+    global _valmgr_abi, _valmgr_code
```

Comment:
I would very much like to avoid the use of globals.  Can we restructure this so that these values are not lazily computed or that it's an error if they are not populated?

Reply:
Is it legible that `create_valmgr_tx` creates the vmc tx every time it is called, and we provide functions `get_abi(vmc_tx)`, `get_addr(vmc_tx)`, `get_sender_addr(vmc_tx)`, ... to get these information from the created vmc tx?
Or I can provide another class like `VMC`, which creates the transaction once it is initialized, and support methods for people to get those information?

Reply:
I'll have to get back to you monday when I have some time to digest this and figure out where a good place to store/generate this stuff would be.